### PR TITLE
Added a different chain for iptables rules

### PR DESCRIPTION
**Pull request by rbrtbnfgl**
## Description

Added a different chain for Flannel iptables rules to order the rules before the one already present on the node.
It should solve #1619 #1542 #1033

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: network/iptables.go [hunk 1]
```diff
@@ -135,7 +173,7 @@ func ipTablesRulesExist(ipt IPTables, rules []IPTablesRule) (bool, error) {
 }
 
 // ipTablesCleanAndBuild create from a list of iptables rules a transaction (as string) for iptables-restore for ordering the rules that effectively running
-func ipTablesCleanAndBuild(ipt IPTables, rules []IPTablesRule) (IPTablesRestoreRules, error) {
+func ipTablesCleanAndBuild(ipt IPTables, rules []IPTablesRule, toChain bool) (IPTablesRestoreRules, error) {
```
Comment (manuelbuil) [inline network/iptables.go:176]: Can you explain in the function comment what "toChain" mean?

File: network/iptables.go [hunk 2]
```diff
@@ -43,6 +43,14 @@ type IPTablesRule struct {
...
 }
 
+var MasqChain = []IPTablesRule{
+	{"nat", "POSTROUTING", []string{"-m", "comment", "--comment", "flanneld masq", "-j", "FLANNEL-POSTRTG"}},
+}
+
+var FwdChain = []IPTablesRule{
+	{"filter", "FORWARD", []string{"-m", "comment", "--comment", "flanneld forward", "-j", "FLANNEL-FWD"}},
+}
+
```
Comment (manuelbuil) [inline network/iptables.go:53]: You can have both definitions in one var ( ...) section

File: network/iptables_windows.go [hunk 3]
```diff
@@ -31,11 +31,21 @@ type IPTablesRule struct {
...
-func MasqIP6Rules(ipn ip.IP6Net, lease *subnet.Lease) []IPTablesRule { return nil }
-func DeleteIP4Tables(rules []IPTablesRule) error                     { return nil }
-func DeleteIP6Tables(rules []IPTablesRule) error                     { return nil }
+var MasqChain = []IPTablesRule{
+	{"nat", "POSTROUTING", []string{"-m", "comment", "--comment", "flanneld masq", "-j", "FLANNEL-POSTRTG"}},
+}
+
+var FwdChain = []IPTablesRule{
+	{"filter", "FORWARD", []string{"-m", "comment", "--comment", "flanneld forward", "-j", "FLANNEL-FWD"}},
+}
```
Comment (manuelbuil) [inline network/iptables_windows.go:40]: You can define both variables in one var section

File: network/iptables_test.go [hunk 4]
```diff
@@ -114,7 +114,7 @@ func TestDeleteRules(t *testing.T) {
 	baseRules := MasqRules(ip.IP4Net{}, lease())
 	expectedRules := expectedTearDownIPTablesRestoreRules(baseRules)
 
-	err := ipTablesBootstrap(ipt, iptr, baseRules)
+	err := ipTablesBootstrap(ipt, iptr, baseRules, false)
```
Comment (manuelbuil) [inline network/iptables_test.go:117]: All use `toChain=false`, is it possible to create one that uses `toChain=true`?

File: network/iptables.go (hunk 1 above)
Comment (rbrtbnfgl) [reply network/iptables.go:176]: If the rule is used to send the traffic to the chain and it has to be added with -I and not -A.
Maybe I can modified to a more meaningful name.

File: network/iptables.go [hunk 5]
```diff
@@ -88,34 +93,70 @@ func MasqIP6Rules(ipn ip.IP6Net, lease *subnet.Lease) []IPTablesRule {
...
+		log.Errorf("Failed to setup IPTables. Error on creating the chain: %v", err)
+		return
+	}
+}
+
+func CreateIP6Chain(table, chain string) {
+	ipt, err := iptables.NewWithProtocol(iptables.ProtocolIPv6)
+	if err != nil {
+		// if we can't find iptables, give up and return
+		log.Errorf("Failed to setup IPTables. iptables binary was not found: %v", err)
```
Comment (manuelbuil) [inline network/iptables.go:152]: Could you add v6 in the logs? Just to differentiate with the ipv4 case. Imagine we get an error in a dual-stack env, we would not know where it is coming from

**Review by manuelbuil — COMMENTED** [review]
Looks good. Only one comment

**Review by manuelbuil — APPROVED** [review]

**Review by thomasferrandiz — APPROVED** [review]

**Pull request merged**